### PR TITLE
RFC: remove the branch dependency of git-externals

### PR DESCRIPTION
**Pull request by tkelman**
~~in order to catch checksum mismatches more quickly~~

~~Ignore the branch name,~~ I'm repurposing this to address https://github.com/JuliaLang/julia/pull/10743#issuecomment-107050253 by removing the branch-tracking misfeature of git-externals. Instead we should always download a specific sha, instead of cloning whatever the tip of `FOO_BRANCH` says and only warning if it doesn't match `FOO_SHA1` from `deps/foo.version`. This will make the dependency builds that are based on git checkouts a little more future-proof (I hope).

We can also probably delete the separate checksum files for these in `deps/checksums`, which only get used when `NO_GIT = 1` is set. They're more cryptographically secure than SHA1 since they use SHA256 when available, but I don't see that as a huge issue. The checksums are really for integrity checking of flaky downloads, not bulletproof security. If someone feels really strongly about this we can keep them.

### CODE REVIEW

File: deps/Makefile [hunk 1]
```diff
@@ -246,11 +246,11 @@ include $1.version
 ifneq ($(NO_GIT),1)
 $2_SRC_DIR = $1
 $1:
-	git clone -q --depth=50 --branch $$($2_BRANCH) $$($2_GIT_URL) $1
+	git clone -q --depth=50 $$($2_GIT_URL) $1
 	touch -c $1/$3
 $1/$3: $1.version | $1
-	cd $1 && git fetch -q $$($2_GIT_URL) $$($2_BRANCH):remotes/origin/$$($2_BRANCH)
-	cd $1 && git checkout -q --detach remotes/origin/$$($2_BRANCH)
+	cd $1 && git fetch -q $$($2_GIT_URL)
```
Comment (tkelman) [inline deps/Makefile]: cc @vtjnash the fact that we're doing a shallow clone may cause problems if we try to change `*.version` to a commit that's on a different branch, I was doing some dummy tests and it wasn't always working as I'd hoped.

**Comment by tkelman** [comment]
To convince myself this works properly I ran `make NO_GIT=1 distcleanall && make distcleanall && make NO_GIT=1 -j4 test && make NO_GIT=1 distcleanall && make distcleanall && make -j4 test` locally.

File: deps/Makefile [hunk 2]
```diff
@@ -263,7 +263,7 @@ endif
 else # NO_GIT
 $2_SRC_DIR = $1-$$($2_BRANCH)
```
Comment (vtjnash) [inline deps/Makefile]: probably best to rename the file to use the SHA1 hash also, so we don't confuse the checksum system.

Comment (tkelman) [inline deps/Makefile]: I suppose so, thanks. What's your opinion, should I rename the checksum files or delete them?

Comment (vtjnash) [inline deps/Makefile]: i suppose it's best to keep them for anyone who does a direct download of the github tarball?

Comment (tkelman) [inline deps/Makefile]: Ok. For keeping them updated I may want to reinstate what this PR originally was doing, which is building with `NO_GIT` on Travis and AppVeyor.

**Comment by tkelman** [comment]
Okay to merge? Anyone have strong feelings on the checksums for git-externals deps, or which of these commits should maybe be squashed together?

**Comment by tkelman** [comment]
I'll take the silence as a "go ahead and merge" then.

**Pull request merged**